### ISSUE TITLE
Issues with local-path (vendored) wheel dependency and pypi using conditional environment markers

### ISSUE THREAD

**Issue by hozn** [python-poetry/poetry#7452]
- **Poetry version**: 1.3.2
- **Python version**: 3.10.2
- **OS version and name**: MacOS/Darwin 13.1, M1 (arm64)
- **pyproject.toml**: https://gist.github.com/hozn/e7c67bcc7a4639fff3d2cd6638d6ace8

- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) stable Poetry version, installed using a recommended method.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have consulted the [FAQ](https://python-poetry.org/docs/faq/) and [blog](https://python-poetry.org/blog/) for any relevant entries or release notes.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option) and have included the output below. (N/A)

### Updated Issue Description
(Based on thread below, updating this to articulate the *actual* problem, which the original did not correctly pinpoint.)

The **order** of the dependencies and whether `source` is specified (or left at default value) seems to be a factor when creating the `poetry.lock` file.

`pyproject.toml` (snippet):

```toml
[tool.poetry.dependencies]
Python = "^3.10"
# <snip>
grpcio = [
    {markers="sys_platform == 'darwin' and platform_machine=='arm64'", path='./vendored/grpcio-1.51.1-cp310-cp310-macosx_11_0_arm64.whl'},
    {version="==1.51.1", markers="sys_platform != 'darwin' or platform_machine != 'arm64'"},
]
```

Generates a `poetry.lock` file that only includes information about the vendored macos-specific gprcio wheel.
Broken poetry.lock file: https://gist.github.com/hozn/c2ae7bbadec0437261a0b92b4fd07508

Ways to fix this:
1. Change the order of the two lines.
2. Or add `source="pypi"` to the non-vendored dependency.

### Original Issue Description
On MacOS M1  (arm64) it is notoriously painful to install grpcio.  For that reason, I need to install from a prebuilt local (vendored) path dependency when the environment markers indicate darwin+arm64.  This is how I'm attempting to achieve this in the `pyproject.toml`:

```toml
[tool.poetry.dependencies]
Python = "^3.10"
# <snip>
grpcio = [
    {version="==1.51.1", markers="sys_platform != 'darwin' or platform_machine != 'arm64'", source="pypi"},
    {markers="sys_platform == 'darwin' and platform_machine=='arm64'", path='./vendored/grpcio-1.51.1-cp310-cp310-macosx_11_0_arm64.whl'},
]
```

1. With the code above, `poetry lock` will generate a lockfile that only includes the package instances / hashes from pypi -- not the locally vendored version.
2. If I remove the `source="pypi"` from the first dep line, running `poetry lock` will generate a lockfile that *only* includes the hash for the vendored dependency.

I have not managed to get poetry to create a `poetry.lock` file that includes hashes for both the vendored package and the pypi packages.

**Comment by dimbleby** [comment]
I see the same lockfile (modulo the `content-hash`) whether or not the source is specified.  

Do you trust your own eyes at this point?!

**Comment by hozn** [comment]
Ok, apparently there is more nuance to this / I had changed more than one variable in attempting to fix this ... 

When I change the *order* of the 2 dependency package sources **as well as** removing the `source="pypi"`, then it breaks.

From:
```toml
grpcio = [
    {version="==1.51.1", markers="sys_platform != 'darwin' or platform_machine != 'arm64'"},
    {markers="sys_platform == 'darwin' and platform_machine=='arm64'", path='./vendored/grpcio-1.51.1-cp310-cp310-macosx_11_0_arm64.whl'},
]
```
to
```toml
grpcio = [
    {markers="sys_platform == 'darwin' and platform_machine=='arm64'", path='./vendored/grpcio-1.51.1-cp310-cp310-macosx_11_0_arm64.whl'},
    {version="==1.51.1", markers="sys_platform != 'darwin' or platform_machine != 'arm64'"},
]
```

Rerunning `poetry lock` after making this change results in removing the other hashes.  Now my grpcio block looks like:
```toml
[[package]]
name = "grpcio"
version = "1.51.1"
description = "HTTP/2-based RPC framework"
category = "main"
optional = false
python-versions = ">=3.7"
files = [
    {file = "grpcio-1.51.1-cp310-cp310-macosx_11_0_arm64.whl", hash = "sha256:d6cd2c5190eb25623fe175a8646f615477786c1118ac7f440591f25e5b66a51b"},
]

[package.extras]
protobuf = ["grpcio-tools (>=1.51.1)"]

[package.source]
type = "file"
url = "vendored/grpcio-1.51.1-cp310-cp310-macosx_11_0_arm64.whl"
```
- Here's the poetry.lock file: https://gist.github.com/hozn/c2ae7bbadec0437261a0b92b4fd07508
- Here's the the breaking pyproject.toml file: https://gist.github.com/hozn/e7c67bcc7a4639fff3d2cd6638d6ace8

@dimbleby , do let me know if your eyes see something else, though.

**Comment by elanou** [comment]
I can't get the platform markers to work either when trying to specify platform specific wheels for torch.

I've tried defining them as 
```
platform = 'darwin' 
markers = "sys_platform == \"linux\""
markers = "sys_platform == 'linux'"
```
example dependency:
```
[tool.poetry.dependencies]
python = "^3.9"

torch = [
    {url = "https://download.pytorch.org/whl/cu116/torch-1.13.1%2Bcu116-cp39-cp39-linux_x86_64.whl",  markers = "sys_platform == \"linux\""},
    {url = "https://download.pytorch.org/whl/cpu/torch-1.13.1-cp39-none-macosx_10_9_x86_64.whl", markers = "sys_platform == \"darwin\""}
]
```

The poetry.lock then has no concept of the platform markers:
```
[[package]]
name = "torch"
version = "1.13.1"
description = "Tensors and Dynamic neural networks in Python with strong GPU acceleration"
category = "main"
optional = false
python-versions = ">=3.7.0"
files = [
    {file = "torch-1.13.1-cp39-none-macosx_10_9_x86_64.whl", hash = "sha256:6930791efa8757cb6974af73d4996b6b50c592882a324b8fb0589c6a9ba2ddaf"},
]

[package.dependencies]
typing-extensions = "*"

[package.extras]
opt-einsum = ["opt-einsum (>=3.3)"]

[package.source]
type = "url"
url = "https://download.pytorch.org/whl/cpu/torch-1.13.1-cp39-none-macosx_10_9_x86_64.whl"
[[package]]
name = "torch"
version = "1.13.1+cu116"
description = "Tensors and Dynamic neural networks in Python with strong GPU acceleration"
category = "main"
optional = false
python-versions = ">=3.7.0"
files = [
    {file = "torch-1.13.1%2Bcu116-cp39-cp39-linux_x86_64.whl", hash = "sha256:db457a822d736013b6ffe509053001bc918bdd78fe68967b605f53984a9afac5"},
]

[package.dependencies]
typing-extensions = "*"

[package.extras]
opt-einsum = ["opt-einsum (>=3.3)"]

[package.source]
type = "url"
url = "https://download.pytorch.org/whl/cu116/torch-1.13.1%2Bcu116-cp39-cp39-linux_x86_64.whl"
```

When trying to install on a linux system it attempts to install the macOS wheel, presumably because its listed first.

**Comment by elanou** [comment]
Update on my situation above: If I go in and manually update the poetry.lock file to have the sys_platform markers then the correct wheel is installed. So for some reason the markers are being excluded when the lock file is generated.

Manually updated lock:
```
[[package]]
name = "torch"
version = "1.13.1"
description = "Tensors and Dynamic neural networks in Python with strong GPU acceleration"
category = "main"
optional = false
python-versions = ">=3.7.0"
files = [
    {file = "torch-1.13.1-cp39-none-macosx_10_9_x86_64.whl", hash = "sha256:6930791efa8757cb6974af73d4996b6b50c592882a324b8fb0589c6a9ba2ddaf", markers = "sys_platform == \"darwin\""},
]

[package.dependencies]
typing-extensions = "*"

[package.extras]
opt-einsum = ["opt-einsum (>=3.3)"]

[package.source]
type = "url"
url = "https://download.pytorch.org/whl/cpu/torch-1.13.1-cp39-none-macosx_10_9_x86_64.whl"
[[package]]
name = "torch"
version = "1.13.1+cu116"
description = "Tensors and Dynamic neural networks in Python with strong GPU acceleration"
category = "main"
optional = false
python-versions = ">=3.7.0"
files = [
    {file = "torch-1.13.1%2Bcu116-cp39-cp39-linux_x86_64.whl", hash = "sha256:db457a822d736013b6ffe509053001bc918bdd78fe68967b605f53984a9afac5", markers = "sys_platform == \"linux\""},
]

[package.dependencies]
typing-extensions = "*"

[package.extras]
opt-einsum = ["opt-einsum (>=3.3)"]

[package.source]
type = "url"
url = "https://download.pytorch.org/whl/cu116/torch-1.13.1%2Bcu116-cp39-cp39-linux_x86_64.whl"
[[package]]
name = "typing-extensions"
version = "4.4.0"
description = "Backported and Experimental Type Hints for Python 3.7+"
category = "main"
optional = false
python-versions = ">=3.7"
files = [
    {file = "typing_extensions-4.4.0-py3-none-any.whl", hash = "sha256:16fa4864408f655d35ec496218b85f79b3437c829e93320c7c9215ccfd92489e"},
    {file = "typing_extensions-4.4.0.tar.gz", hash = "sha256:1511434bb92bf8dd198c12b1cc812e800d4181cfcb867674e0f8279cc93087aa"},
]
```

**Comment by dimbleby** [comment]
> Ways to fix this:

> 1.   Change the order of the two lines.
> 2.   Or add source="pypi" to the non-vendored dependency.

2 is I think the right answer, and this all is working correctly.

See also #7722, which basically duplicates this

**Comment by ralbertazzi** [comment]
@elanou the use case with multiple url dependencies seems to be fixed already in 1.5. As far as I see the problem appears when a "direct origin" dependency (url, git, directory, ..) and a "source" dependency (pypi, private source) are mixed together.

Also - I know it doesn't strictly fix the issue but just FYI - M1 wheels for grpcio have been released a couple months ago

**Comment by ralbertazzi** [comment]
I confirm this has no problems on Poetry 1.5. I modified the wheels to speed up download and since I'm on Mac M1:
```pyhon
torch = [
    { url = "https://download.pytorch.org/whl/cpu/torch-1.13.1-cp39-none-macosx_11_0_arm64.whl", markers = "sys_platform == \"darwin\"" },
    { url = "https://download.pytorch.org/whl/cpu/torch-1.13.1%2Bcpu-cp39-cp39-linux_x86_64.whl", markers = "sys_platform == \"linux\"" },
]
```
